### PR TITLE
eth/downloader: fewer headers and futures too in ancestor lookup

### PR DESCRIPTION
**Pull request by karalabe**
*This is a followup PR to https://github.com/ethereum/go-ethereum/pull/2861, please merge that in first.*

Although we strive to maintain an accurate remote chain head metadata about all of our connected peers, for new connections we kind of have to take in on face value that the peer is truthful in the handshake (especially if it's in the future, there's no sure way to verify).

However there are some peers in the network which report high TDs in their handshakes, but low head numbers. This causes problems for the sync algo as the high TD will prompt a sync cycle, but the low reported head causes the ancestor lookup to estimate an older fork point, pulling and reimporting many duplicate blocks (wasted IO, wasted PoW check).

This PR modified the ancestor lookup so that instead of requesting the top 192 headers down from `H = min(our-head, remote-head)`, it requests `[H - 192, H + 192)` with 15 header gaps (i.e. in total 24 headers, 12 in the past, 1 advertised head and 11 in the future). If the peer had some anomaly in not reporting its best head, just an approximate, then we'll notice this and can actually sync higher up from the chain.

If however the peer satisfied out entire query (i.e. all 24 headers) and we actually knew all of them, it means the peer lied and tried to stall us / waste our resources. In that case drop the peer off.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Jul 26 08:31:51 UTC 2016_

**Comment by hdiedrich** [comment]
Thanks, is this supposed to address #2841?

**Comment by karalabe** [comment]
This along with the remaining downloader PRs currently open and being reviewed by @fjl .

File: eth/downloader/downloader.go [hunk 1]
```diff
@@ -587,25 +593,27 @@ func (d *Downloader) findAncestor(p *peer, height uint64) (uint64, error) {
...
+				if headers[i].Number.Int64() < from || headers[i].Number.Uint64() > ceil {
 					continue
 				}
 				// Otherwise check if we already know the header or not
 				if (d.mode == FullSync && d.hasBlockAndState(headers[i].Hash())) || (d.mode != FullSync && d.hasHeader(headers[i].Hash())) {
 					number, hash = headers[i].Number.Uint64(), headers[i].Hash()
+
+					// If every header is known, even future ones, the peer straight out lied about its head
+					if number > height && i == limit-1 {
+						glog.V(logger.Warn).Infof("%v: lied about chain head: reported %d, found above %d", p, height, number)
```
Comment (obscuren) [inline eth/downloader/downloader.go]: [Why you always lying](https://www.youtube.com/watch?v=dsswAbB725o)

**Comment by obscuren** [comment]
From a code POV it looks good (:+1), though I have a hard time understanding the implications this could have.

File: eth/downloader/downloader.go (hunk 1 above)
Comment (karalabe) [inline eth/downloader/downloader.go]: lol :) nice one

**Comment by obscuren** [comment]
Pling @robotally

**Pull request merged**